File: src/types/trips.ts
import type { User } from './auth.ts';
import type { Dayjs } from 'dayjs';
import type { RecordModel } from 'pocketbase';

export type Entity = Omit<RecordModel, 'collectionId' | 'collectionName'>;

export type Participant = {
  name: string;
  email?: string;
  userId?: string;
};

export type Place = Entity & {
  stateName?: string;
  countryName?: string;
  latitude?: string;
  longitude?: string;
  name: string;
  category?: string;
  timezone?: string;
};

export type Trip = Entity & {
  ownerId: string;
  name: string;
  description?: string;
  notes?: string;
  startDate: string;
  endDate: string;
  coverImage?: string;
  participants?: Participant[];
  destinations?: Place[];
  collaborators?: User[];
  budget?: Cost;
};

export type NewTrip = Omit<Trip, 'id'>;

export type CreateTripForm = {
  name: string;
  description?: string;
  dateRange: [string | null, string | null];
  coverImage?: string;
  participants?: string[];
  destinations?: Place[];
  budgetAmount?: number;
  budgetCurrency?: string;
};

export type Cost = {
  value: number;
  currency: string;
};

export type Attachment = {
  id: string;
  name: string;
  file: string;
};

export type Expense = Entity & {
  name: string;
  trip: string;
  cost?: Cost;
  occurredOn?: string;
  notes?: string;
  category?: string;
  attachmentReferences?: string[];
};

export type ConvertedExpense = Expense & {
  convertedCost?: Cost;
};

export type CreateExpense = Omit<Expense, 'id'>;

export type Transportation = Entity & {
  type: string;
  origin: string;
  destination: string;
  cost: Cost;
  departureTime: string;
  arrivalTime: string;
  trip: string;
  metadata: { [key: string]: any };
  attachments?: string[];
  attachmentReferences?: string[];
  expenseId?: string;
};

export type CreateTransportation = {
  type: string;
  origin: string;
  destination: string;
  cost?: Cost;
  departureTime: string;
  arrivalTime: string;
  trip: string;
  metadata?: { [key: string]: any };
  attachments?: string[];
  attachmentReferences?: string[];
  expenseId?: string;
  link?: string;
};

export type CarRentalFormSchema = {
  rentalCompany?: string;
  pickupLocation?: string;
  dropOffLocation?: string;
  pickupTime?: string;
  dropOffTime?: string;
  confirmationCode?: string;
  cost?: number;
  currencyCode?: string;
  link?: string;
  place?: Place;
};

export type TransportationFormSchema = {
  provider?: string;
  origin?: string;
  destination?: string;
  departureTime?: string;
  arrivalTime?: string;
  reservation?: string;
  cost?: number;
  currencyCode?: string;
  originAddress?: string;
  destinationAddress?: string;
  flightNumber?: string;
  assignedSeats?: string;
  link?: string;
};

export type FlightFormSchema = Omit<TransportationFormSchema, 'origin' | 'destination'> & {
  flightNumber?: string;
  seats?: string;
  origin?: Airport;
  destination: Airport;
};

export type BikeForSchema = TransportationFormSchema & {
  elevation?: number;
  distance?: string;
};

export type CroppedImage = {
  height: number;
  width: number;
  x: number;
  y: number;
};

export type Collaborator = {
  id?: string;
  trip: string;
  user: string | User;
};

export type Lodging = {
  id: string;
  type: string;
  name: string;
  address?: string;
  cost?: Cost;
  startDate: string;
  endDate: string;
  trip: string;
  confirmationCode?: string;
  metadata?: { [key: string]: any };
  attachments?: string[];
  attachmentReferences?: string[];
  expenseId?: string;
  link?: string;
};

export type CreateLodging = Omit<Lodging, 'id'>;

export type LodgingFormSchema = {
  type?: string;
  name?: string;
  address?: string;
  cost?: number;
  currencyCode?: string;
  startDate?: string;
  endDate?: string;
  confirmationCode?: string;
  place?: Place;
  link?: string;
};

export const enum LodgingType {
  HOTEL = 'hotel',
  HOME = 'home',
  RENTAL = 'vacation_rental',
  CAMP_SITE = 'camp_site',
}

export interface Activity extends RecordModel {
  id: string;
  name: string;
  description: string;
  address?: string;
  startDate: string;
  endDate?: string;
  cost?: Cost;
  trip: string;
  attachments?: string[];
  attachmentReferences?: string[];
  expenseId?: string;
}

export type CreateActivity = Omit<Activity, 'id'>;

export type ActivityFormSchema = {
  name?: string;
  description?: string;
  address?: string;
  cost?: number;
  currencyCode?: string;
  startDate?: string;
  endDate?: string;
  place?: Place;
  link?: string;
};

export interface Airport extends Omit<RecordModel, 'collectionName,collectionId'> {
  iataCode?: string;
  name: string;
}

export interface Airline extends Omit<RecordModel, 'collectionName,collectionId'> {
  code?: string;
  logo?: string;
  name: string;
}

export type ItineraryLine = (Transportation | Lodging | Activity) & {
  itineraryType?: string;
  day: Dayjs;
};

File: backend/types/trips.go
package types

import "github.com/pocketbase/pocketbase/tools/types"

type VersionInfo struct {
	Tag    string `json:"tag"`
	Commit string `json:"commit"`
	Dirty  bool   `json:"dirty"`
}

type UploadedFile struct {
	Id          string `json:"id"`
	FileName    string `json:"fileName"`
	FileContent string `json:"fileContent"`
}

type Attachment struct {
	Id   string `json:"id"`
	Name string `json:"name"`
	File string `json:"file"`
}

type Destination struct {
	Id          string `json:"id"`
	Name        string `json:"name"`
	StateName   string `json:"stateName"`
	CountryName string `json:"countryName"`
	TimeZone    string `json:"timezone"`
	Latitude    string `json:"latitude"`
	Longitude   string `json:"longitude"`
}

type Participant struct {
	Name string `json:"name"`
}

type Cost struct {
	Value    float64 `json:"value"`
	Currency string  `json:"currency"`
}

type Transportation struct {
	Id                   string          `json:"id"`
	Type                 string          `json:"type"`
	Origin               string          `json:"origin"`
	Destination          string          `json:"destination"`
	Cost                 *Cost           `json:"cost"`
	Departure            types.DateTime  `json:"departure"`
	Arrival              types.DateTime  `json:"arrival"`
	Attachments          []*UploadedFile `json:"attachments"`
	AttachmentReferences []string        `json:"attachmentReferences"`
	Metadata             map[string]any  `json:"metadata"`
}

type Lodging struct {
	Id                   string          `json:"id"`
	Type                 string          `json:"type"`
	Name                 string          `json:"name"`
	Address              string          `json:"address"`
	ConfirmationCode     string          `json:"confirmationCode"`
	Cost                 *Cost           `json:"cost"`
	StartDate            types.DateTime  `json:"startDate"`
	EndDate              types.DateTime  `json:"endDate"`
	Attachments          []*UploadedFile `json:"attachments"`
	AttachmentReferences []string        `json:"attachmentReferences"`
	Metadata             map[string]any  `json:"metadata"`
}

type Activity struct {
	Id                   string          `json:"id"`
	Name                 string          `json:"name"`
	Description          string          `json:"description"`
	Address              string          `json:"address"`
	ConfirmationCode     string          `json:"confirmationCode"`
	Cost                 *Cost           `json:"cost"`
	StartDate            types.DateTime  `json:"startDate"`
	EndDate              types.DateTime  `json:"endDate"`
	Attachments          []*UploadedFile `json:"attachments"`
	AttachmentReferences []string        `json:"attachmentReferences"`
	Metadata             map[string]any  `json:"metadata"`
}

type Expense struct {
	Id                   string         `json:"id"`
	Name                 string         `json:"name"`
	Cost                 *Cost          `json:"cost"`
	OccurredOn           types.DateTime `json:"occurredOn"`
	Notes                string         `json:"notes"`
	Category             string         `json:"category"`
	AttachmentReferences []string       `json:"attachmentReferences"`
}

type Trip struct {
	Id                 string         `json:"id"`
	Name               string         `json:"name"`
	Description        string         `json:"description"`
	StartDate          types.DateTime `json:"startDate"`
	EndDate            types.DateTime `json:"endDate"`
	Destinations       []Destination  `json:"destinations"`
	Participants       []Participant  `json:"participants"`
	CoverImage         *UploadedFile  `json:"coverImage"`
	CoverImageFileName string         `json:"coverImageFileName"`
	Notes              string         `json:"notes"`
	Budget             *Cost          `json:"budget"`
}

type ExportedTrip struct {
	Trip            *Trip             `json:"trip"`
	Transportations []*Transportation `json:"transportations"`
	Lodgings        []*Lodging        `json:"lodgings"`
	Activities      []*Activity       `json:"activities"`
	Expenses        []*Expense        `json:"expenses"`
	Attachments     []*Attachment     `json:"attachments"`
}

type Airport struct {
	Id         string `json:"id"`
	Name       string `json:"name"`
	Latitude   string `json:"latitude"`
	Longitude  string `json:"longitude"`
	Timezone   string `json:"timezone"`
	IataCode   string `json:"iataCode"`
	IsoCountry string `json:"isoCountry"`
}

type Airline struct {
	Id   string `json:"id"`
	Name string `json:"name"`
}
